diff --git a/dmenu.c b/dmenu.c
index 8acf057..ed4bbbf 100644
--- a/dmenu.c
+++ b/dmenu.c
@@ -39,6 +39,7 @@ static void setup(void);
 static void usage(void);
 
 static char text[BUFSIZ] = "";
+static char originaltext[BUFSIZ] = "";
 static int bh, mw, mh;
 static int inputw, promptw;
 static int lines = 0;
@@ -369,10 +370,43 @@ keypress(XKeyEvent *ev) {
 	case XK_Tab:
 		if(!sel)
 			return;
-		strncpy(text, sel->text, sizeof text);
-		cursor = strlen(text);
-		match(True);
+        if(strcmp(text, sel->text)) {
+            strncpy(originaltext, text, sizeof originaltext);
+		    strncpy(text, sel->text, sizeof text);
+		    cursor = strlen(text);
+        } else {
+            if(sel && sel->right) {
+                sel = sel->right;
+		        strncpy(text, sel->text, sizeof text);
+		        cursor = strlen(text);
+            }
+            else {
+                strncpy(text, originaltext, sizeof text);
+                cursor = strlen(text);
+                match(True);
+            }
+        }
 		break;
+    case XK_ISO_Left_Tab:
+        if(!sel)
+			return;
+        if(strcmp(text, sel->text)) {
+            sel = matchend;
+            strncpy(originaltext, text, sizeof originaltext);
+		    strncpy(text, sel->text, sizeof text);
+		    cursor = strlen(text);
+        } else {
+            if(sel && sel->left) {
+                sel = sel->left;
+		        strncpy(text, sel->text, sizeof text);
+		        cursor = strlen(text);
+            }
+            else {
+                strncpy(text, originaltext, sizeof text);
+                cursor = strlen(text);
+                match(True);
+            }
+        } 
 	}
 	drawmenu();
 }
